protomap/proto.go

```diff
@@ -18,63 +18,361 @@
+			for k, v := range pl.mappedValues {
+				mappedValues[k] = v
+			}
+
+			// If this was the list value, then we return the value of the list,
+			// which is always a protobuf message back.
+			if pl.member != nil {
+				listVal = pl.member
+			}
+			return true
+		})
+
+		if listParseErr != nil {
+			return fmt.Errorf("could not parse a field within the list %s , %v", fd.FullName(), listParseErr)
+		}
+
+		// Handle recursing down the tree for the list that we just extracted.
+		p := proto.Clone(listPath).(*gpb.Path)
+		if basePath != nil {
+			// This is the first time that we have found a path that requires a
+			// data tree path, not a schema tree path.
+			p = resolvedPath(basePath, listPath)
+		}
+
+		for kn, kv := range listKeys {
+			le := p.Elem[len(p.Elem)-1]
+			if le.Key == nil {
+				le.Key = map[string]string{}
+			}
+			le.Key[kn] = kv
+		}
+
+		for path, value := range mappedValues {
+			vals[resolvedPath(p, path)] = value
+		}
+
+		if err := pathsFromProtoInternal(listVal, vals, p); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// parsedListField returns the details of an individual field of a message
+// which corresponds to a YANG list (is 'repeated').
+type parsedListField struct {
+	// keys is a map of the keys using the gNMI path format for keys - where all
+	// values are mapped to strings. Since the parsed field contains only one field
+	// then only one key will ever be populated. This field is only set to a non-nil
+	// value if the field corresponds to a key.
+	keys map[string]string
+	// mappedValues stores the values that are contained in the input protobuf, and is
+	// populated only when the field supplied is a key. Since the key fields are removed
+	// from the 'value' protobuf, their values can only be extracted from the repeated 'Key'
+	// message. A single field may result in >1 populated mappedValues in the case that there
+	// are multiple paths within the schemapath annotation.
+	mappedValues map[*gpb.Path]interface{}
+	// member is populated when the field parsed is the member of the list - i.e., the
+	// repeated proto.Message which corresponds to the subtree under the list at a particular
+	// key.
+	member proto.Message
+}
+
+// parseListField parses an individual field within a 'repeated' message representing a YANG list, as
+// described by fd, and with value v. The supplied basePath is used for the base data tree path for field
+// specified. It returns a parsedListField describing the individual field supplied.
+func parseListField(fd protoreflect.FieldDescriptor, v protoreflect.Value, basePath *gpb.Path) (*parsedListField, error) {
+	if fd.IsMap() || fd.IsList() {
+		return nil, fmt.Errorf("invalid field in list, %s", fd.FullName())
+	}
+
+	if fd.Kind() == protoreflect.MessageKind {
+		if t, ok := v.Message().Interface().(proto.Message); ok {
+			// The only case of having proto.Message in a list key is when it is the
+			// value itself, therefore return this value.
```

Comment:
I feel like "the value itself" requires too much context to make sense of -- I thought it meant the entire list value itself at first. Maybe "the message containing the list's value portion" is better?